### PR TITLE
fix(v4): Fix panic when forming contracts with different basis than the host

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
Fixes https://github.com/SiaFoundation/coreutils/issues/143

### CODE REVIEW

File: rhp/v4/server.go [hunk 1]
```diff
@@ -585,14 +585,14 @@ func (s *Server) handleRPCFormContract(stream net.Conn) error {
 
 	// update renter input basis to reflect our funding basis
 	if basis != req.Basis {
-		hostInputs := formationTxn.SiacoinInputs[len(formationTxn.SiacoinInputs)-len(req.RenterInputs)]
-		formationTxn.SiacoinInputs = formationTxn.SiacoinInputs[:len(formationTxn.SiacoinInputs)-len(req.RenterInputs)]
+		hostInputs := formationTxn.SiacoinInputs[len(req.RenterInputs):]
```
Comment (n8maninger) [inline rhp/v4/server.go:588]: This will likely need to be fixed in RPC refresh and renew

Comment (ChrisSchinnerl) [reply rhp/v4/server.go:588]: Yeah I know. Already working on it.

File: rhp/v4/rpc_test.go [hunk 2]
```diff
@@ -266,10 +267,44 @@ func TestFormContractBasis(t *testing.T) {
...
+	cm, s, _ := startTestNode(t, n, genesis)
+
+	// create a separate wallet that doesn't automatically sync with the test node
+	wCS, wState, err := chain.NewDBStore(chain.NewMemDB(), n, genesis)
+	if err != nil {
+		t.Fatal(err)
+	}
+	wCM := chain.NewManager(wCS, wState)
+	wStore := testutil.NewEphemeralWalletStore()
+	w, err := wallet.NewSingleAddressWallet(renterKey, wCM, wStore)
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply rhp/v4/rpc_test.go:279]: Seems like they do cause the test panics without the changes to the handler

Comment (n8maninger) [inline rhp/v4/rpc_test.go:279]: This whole bit can probably still use `startTestNode`, just don't connect the syncs

File: rhp/v4/server.go [hunk 3]
```diff
@@ -585,14 +585,14 @@ func (s *Server) handleRPCFormContract(stream net.Conn) error {
...
+		formationTxn.SiacoinInputs = formationTxn.SiacoinInputs[:len(req.RenterInputs)]
 		txnset, err := s.chain.UpdateV2TransactionSet([]types.V2Transaction{formationTxn}, req.Basis, basis)
 		if err != nil {
 			return errorBadRequest("failed to update renter inputs from %q to %q: %v", req.Basis, basis, err)
 		}
 		formationTxn = txnset[0]
-		formationTxn.SiacoinInputs = append(formationTxn.SiacoinInputs, hostInputs)
+		formationTxn.SiacoinInputs = append(formationTxn.SiacoinInputs, hostInputs...)
```
Comment (lukechampine) [inline rhp/v4/server.go:595]: what happens if we call `UpdateV2TransactionSet` only all the inputs, i.e. without modifying `formationTxn`?

Comment (n8maninger) [reply rhp/v4/server.go:595]: Is it safe to call update on a transaction where some of the inputs are already updated?

Comment (lukechampine) [reply rhp/v4/server.go:595]: hmm, yeah probably not

**Review by n8maninger — APPROVED** [review]

**Pull request merged**